A few cleanups for AsmPrinter

yln opened the pull request:
* Extract common code from `AsmPrinter::GetPOTIndex` overloads
* Don't emit placeholder `llvm.pot` global variable, instead do `if (TM.isPagerando()) EmitPOT();`

Commit:
Don't use else-if after return

Change-Id: I48450f0f120f0ecb8c9a291e7aec30762eb20f1d

Commit:
Extract common AsmPrinter::GetPOTIndex into it's own function

Change-Id: I7ee741471b46e9eef44e0db619dd6b04b762880e

Commit:
Get rid of one of the POT placeholder global variables

Change-Id: I023b61e658191775cb7a6b07381deaea7a26f268

Commit:
Don't actually emit the placeholder

Change-Id: I248347f77dc047efd244760b2c255f71ee59fa79

Commit:
Remove uses of the GV parameter

Change-Id: I2702d2f77d2a9f8147bfc8c8ab63f2f0a8b33526

Commit:
Cleanup EmitPOT

Change-Id: I2ac925559bd708a53203ad87c9fabaa6795955b3

Commit:
Remove StringRef (conversion is implicit)

Change-Id: Ie0341484573a43deec8058cf66c3b5fa77973d67

Commit:
Don't rely on `llvm.pot` placeholder variable to emit POT

Change-Id: I6b6f031607ce4cb203dd6450300a4f4bd41d76f3

Commit:
Don't need to create placeholder variable in PagerandoWrapper pass anymore

Change-Id: I9011c44234879901ade41caf7b119d3333a98168

Commit:
Removal of createPOT allowed additional simplicifactions

Change-Id: I254f3c23930aed648f07325e0df4ff42d703723e

Commit:
Align comment format

Change-Id: Iafcaa1ce11e3ac973d1efb7b3d4f971ec71b4b33

yln commented on lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1668,43 +1669,33 @@ bool AsmPrinter::EmitSpecialLLVMGlobal(const GlobalVariable *GV) {
+  OutStreamer->SwitchSection(Section);
 
+  // Emit POT start label
+  MCSymbol *POTSym = OutContext.getOrCreateSymbol("_POT_");
```
Are global linkage and protected visibility the right choices here?

rinon replied:
Good question. I was using external+protected mostly for ease in debugging. I think we should instead use internal linkage, which implies default visibility. That means we can omit both the `EmitSymbolAttribute` calls.

rinon commented on lib/Target/AArch64/AArch64MCInstLower.cpp:
```diff
@@ -160,7 +160,8 @@ MCOperand AArch64MCInstLower::lowerSymbolOperandELF(const MachineOperand &MO,
     }
     return MCOperand::createImm(index);
-  } else if (SourceFlag == AArch64II::MO_SEC) {
+  }
```
These flags are mutually exclusive. I think it's better to keep the `else` since it's easy to forget to put it back if someone removes the `return`.

rinon commented on include/llvm/CodeGen/AsmPrinter.h:
```diff
@@ -437,13 +437,12 @@ class AsmPrinter : public MachineFunctionPass {
+  /// Return the POT index of the section containing this constant pool ID.
   unsigned GetPOTIndex(unsigned CPID);
+  /// Return the POT index of the section.
+  unsigned GetPOTIndex(const MCSection *Sec);
```
Combining these implementations is great, was thinking the same thing. I think it might make sense to make the `MCSection` form private, since it's only used internally.

rinon reviewed (commented):
Just a couple comments, other than that looks good.

Commit:
Add back `if else` before return

Change-Id: Ibf4a1ce04179a2d89d500ff90fb697155f671d4f

Commit:
Make GetPOTIndex overload private

Change-Id: I1a3a222050da5e3dd7df4e9a9b43b16a5b496a9e

Commit:
Use internal linkage and private visibility

These values are the defaults so we don't need to emit any
attributes for the labels.

Change-Id: I4f73fbec6f4cc1d21789beb1f15fddfb834ba9d9

Commit:
Use EmitReference helper function

Change-Id: I87acca9de29973d0278ab46c2bab234bc76e40dc

Commit:
Rename symbol `_POT_` -> `_PAGE_OFFSET_TABLE_` to be consistent ...

... with `_GLOBAL_OFFSET_TABLE_`

Change-Id: I367d5671cd3f2622eb9317824aa0a2baf8a35cc0

rinon commented on include/llvm/CodeGen/AsmPrinter.h:
```diff
@@ -634,8 +632,9 @@ class AsmPrinter : public MachineFunctionPass {
   /// Emit GlobalAlias or GlobalIFunc.
-  void emitGlobalIndirectSymbol(Module &M,
-                                const GlobalIndirectSymbol& GIS);
+  void emitGlobalIndirectSymbol(Module &M, const GlobalIndirectSymbol &GIS);
```
It's generally frowned on to modify other surrounding lines, even when they're incorrectly formatted. It requires stepping through an extra commit in `git blame`. Not a big deal at all though.

yln replied:
Noted for future changes!

rinon reviewed (commented):
LGTM (with one small comment)
